### ISSUE TITLE
Clone in scope API

### ISSUE THREAD

**Issue by overlookmotel** [oxc-project/oxc#4804]
#4732 introduced `CloneIn` trait.

Currently it clones `SymbolId`, `ScopeId` and `ReferenceId` along with all other parts of AST being cloned.

I don't think this is correct behavior, but am not sure what the correct behavior should be.

If you clone part of an AST which has been through `Semantic`, and insert it elsewhere in same AST, that AST will now contain duplicate IDs.

If you insert it into another AST, the IDs will be invalid. It can cause panic if try to look up the symbol for a `SymbolId`, if the destination AST has less symbols than the source AST, so `SymbolId` is out of bounds. And if it doesn't panic, it'll be the wrong symbol.

Ideally we should provide another API where you provide the destination AST's `ScopeTree` and `SymbolTable`, and what scope the cloned AST fragment will be inserted into, and then it can create new scopes/symbols/references, and patch up all the IDs in the clone automatically for you.

```rs
trait CloneInScope<'a>: Sized {
    type Cloned;
    fn clone_in_scope(
        &self,
        parent_scope_id: ScopeId,
        alloc: &'a Allocator,
        scopes: &mut ScopeTree,
        symbols: &mut SymbolTable,
    ) -> Self::Cloned;
}
```

But in meantime, to avoid invalid IDs, perhaps `CloneIn` should set `scope_id`, `symbol_id` and `reference_id` fields to `None`?

**Comment by hyf0** [comment]
I guess this is also related to the semantics of `PartialEq, Eq, Hash` ... traits on `SymbolId, ScopeId, ReferenceId`.

When you do a comparison between two AST nodes, mostly you don't want to include the `SymbolId, ScopeId, ReferenceId` part.

**Comment by rzvxa** [comment]
👍It is a good idea, I like it. At the very least we should unset those IDs.


I like to know your thoughts on this API. I'm not super familiar with the needs we have for scope tracking so it might be absolute rubbish.

Use this over a `CloneInScope` trait:

```rust
semantic.clone_from(node, scope_id);
```

Then in `Semantic::clone_from` we can clone the node and set its semantic information to turn it into a valid node. Here's how it would happen.

```rust
// where NodeT: CloneIn<T: Scoped>?
let cloned = node.clone_in(alloc); // all IDs are None.
// ...
self.scope_foreign_node(cloned, parent_scope_id); // sets IDs and updates scope via the trait `Scoped`?
```

We can also expose an alias of `scope_foreign_node` to allow resetting semantic data for nodes already in the tree.

The advantage this has over `CloneInScope` is:
* being more flexible.
* We can share the same API with some parts of how we track scope information between runs
* Less maintenance burden
* It also comes with free "separation of concerns" out-of-the-box
* With this, we probably end up with less generated code

The downside is:
* We need an additional trait `Scoped` to be generated - in oxc_semantic - for all of our nodes, However, it can be reused for semantic tracking.
* Leans more toward generics and abstraction which might make the code harder to follow
* `semantic.clone_from(node)` feel less intuitive than `node.clone_in(semantic)`
* You need to have access to the `Semantic` to do this process unless you do it lazily

If you don't have semantics at all, It means you don't care for the semantic information and they should be unset anyway, In those cases, we directly use the `CloneIn::clone_in` method of that node.

**Comment by overlookmotel** [comment]
> When you do a comparison between two AST nodes, mostly you don't want to include the `SymbolId`, `ScopeId`, `ReferenceId` part.

Currently we don't implement `PartialEq` on AST types, but yes the idea is that it should ignore those IDs (and also ignore spans). `Hash`, which is implemented on AST types, does ignore IDs and spans.

**Comment by overlookmotel** [comment]
I'm not sure what exact API or implementation should be for "clone into scope". The APIs you've suggested are better than what I had.

But, at present we couldn't implement it on `Semantic`, because some likely users of this API (transformer, minifier) don't have a complete `Semantic`, only `ScopeTree` and `SymbolTable`.

Maybe that'll change, though - I have some vague ideas as to how to make `Traverse` be able to hold an "entire" `Semantic` including `Nodes`, without violating aliasing rules. Or we could wrap `ScopeTree` and `SymbolTable` into a single struct `Scoping` (like a `Nodes`-less `Semantic`) - I can't think of any occasions where you need one and not the other.

Either way, we'd have a struct which `clone_from` could be a method on.

Anyway, my point at this stage wasn't to work out exact implementation, but just to discuss:

1. Do we need a "clone into scope"-type ability?
2. Would that be feasible to do? (looks like answer is yes)
3. What should `CloneIn` do with IDs?

**Comment by rzvxa** [comment]
Yeah, I was just suggesting an API as pseudo code to work out how we can approach it. You are absolutely right.

> (the reason I ask (1) is that I don't know to what CloneIn is going to be used for, and whether those use cases are operating on a semantic-ified AST or not)

I feel it should unset them, Whatever the use case; having them cloned is unsafe behavior.
Those fields should be behind a feature flag anyway so if you have those in the type it means you are using semantics. Where semantic is enabled, As of now we still can't clone the whole semantic-ified AST since the `AstNode` inner references break.

However, When a time comes that we want to add semantic-ified AST cloning we should add an unsafe method to `CloneIn` like this:

```rust
trait CloneIn<'alloc> {
    fn clone_in(...) -> ...
    
   #[cfg(feature = "semantic")]
   unsafe fn clone_raw_unchecked(...) -> ...
}
```

We can use this method to clone nodes unsafely with their current IDs in places where we know the ID expectations are satisfied.

**Comment by overlookmotel** [comment]
> I feel it should unset them, Whatever the use case; having them cloned is unsafe behavior.

Agree. Have opened #4809 for that change.

We can continue discussing "clone in scope" API on this issue.

**Comment by overlookmotel** [comment]
Transferred to backlog: https://github.com/oxc-project/backlog/issues/127
